File: service/src/main/java/coding/challenge/core/exception/ServiceException.java
package coding.challenge.core.exception;

import lombok.EqualsAndHashCode;
import lombok.Value;

/**
 * Defines an exception that is thrown by service methods.
 */
@Value
@EqualsAndHashCode(callSuper = true)
public class ServiceException extends RuntimeException {

    /**
     * The code of the error.
     */
    int errorCode;

    /**
     * Constructs a new service exception with the specified error code and detail message.
     *
     * @param errorCode The code of the error
     * @param message   The detail message
     */
    public ServiceException(int errorCode, String message) {
        super(message);
        this.errorCode = errorCode;
    }

    /**
     * Constructs a new service exception with the specified error code, detail message and cause.
     *
     * @param errorCode The code of the error
     * @param message   The detail message
     * @param cause     The cause
     */
    public ServiceException(int errorCode, String message, Throwable cause) {
        super(message, cause);
        this.errorCode = errorCode;
    }

}

File: service-kt/src/main/kotlin/coding/challenge/core/exception/ServiceException.kt
package coding.challenge.core.exception

/**
 * Defines an exception that is thrown by service methods.
 */
class ServiceException : RuntimeException {

    /**
     * The code of the error.
     */
    var errorCode: Int

    /**
     * Constructs a new service exception with the specified error code and detail message.
     *
     * @param errorCode The code of the error
     * @param message   The detail message
     */
    constructor(errorCode: Int, message: String?) : super(message) {
        this.errorCode = errorCode
    }

    /**
     * Constructs a new service exception with the specified error code, detail message and cause.
     *
     * @param errorCode The code of the error
     * @param message   The detail message
     * @param cause     The cause
     */
    constructor(errorCode: Int, message: String?, cause: Throwable) : super(message, cause) {
        this.errorCode = errorCode
    }

}
